File: src/api/runtimes/runtimes.ts
/*
 * Copyright (c) 2023-2025 Datalayer, Inc.
 * Distributed under the terms of the Modified BSD License.
 */

/**
 * Runtime instances API functions for the Datalayer platform.
 *
 * Provides functions for managing runtime instances (active compute containers).
 *
 * @module api/runtimes/runtimes
 */

import { requestDatalayerAPI } from '../DatalayerApi';
import { API_BASE_PATHS, DEFAULT_SERVICE_URLS } from '../constants';
import {
  RuntimeData,
  CreateRuntimeRequest,
  CreateRuntimeResponse,
  ListRuntimesResponse,
} from '../../models/RuntimeDTO';
import { validateToken, validateRequiredString } from '../utils/validation';

/**
 * Create a new runtime instance.
 * @param token - Authentication token
 * @param data - Runtime creation configuration
 * @param baseUrl - Base URL for the API (defaults to production Runtimes URL)
 * @returns Promise resolving to the created runtime details
 * @throws {Error} If authentication token is missing or invalid
 * @throws {Error} With status 404 if the environment is not found
 * @throws {Error} With status 503 if no runtime is available
 */
export const createRuntime = async (
  token: string,
  data: CreateRuntimeRequest,
  baseUrl: string = DEFAULT_SERVICE_URLS.RUNTIMES,
): Promise<CreateRuntimeResponse> => {
  validateToken(token);

  try {
    return await requestDatalayerAPI<CreateRuntimeResponse>({
      url: `${baseUrl}${API_BASE_PATHS.RUNTIMES}/runtimes`,
      method: 'POST',
      body: data,
      token,
    });
  } catch (error: any) {
    // Handle specific error cases
    if (error.response) {
      const status = error.response.status;
      const responseData = error.response.data || {};

      if (status === 404) {
        // Environment not found
        throw new Error(
          `Environment '${data.environment_name}' not found. ${responseData.message || 'Please check the environment name and try again.'}`,
        );
      } else if (status === 503) {
        // No runtime available
        throw new Error(
          `No runtime available. ${responseData.message || 'The service is temporarily unavailable or at capacity. Please try again later.'}`,
        );
      }
    }

    // Re-throw the original error for other cases
    throw error;
  }
};

/**
 * List all runtime instances.
 * @param token - Authentication token
 * @param baseUrl - Base URL for the API (defaults to production Runtimes URL)
 * @returns Promise resolving to list of runtime instances
 * @throws {Error} If authentication token is missing or invalid
 */
export const listRuntimes = async (
  token: string,
  baseUrl: string = DEFAULT_SERVICE_URLS.RUNTIMES,
): Promise<ListRuntimesResponse> => {
  validateToken(token);

  const response = await requestDatalayerAPI<any>({
    url: `${baseUrl}${API_BASE_PATHS.RUNTIMES}/runtimes`,
    method: 'GET',
    token,
  });

  // The API returns { success: true, message: string, runtimes: Runtime[] }
  // The response already has the correct structure, just return it
  return response as ListRuntimesResponse;
};

/**
 * Get details for a specific runtime instance.
 * @param token - Authentication token
 * @param podName - The unique pod name of the runtime
 * @param baseUrl - Base URL for the API (defaults to production Runtimes URL)
 * @returns Promise resolving to runtime details
 * @throws {Error} If authentication token is missing or invalid
 * @throws {Error} If pod name is missing or invalid
 * @throws {Error} With status 404 if the runtime is not found
 */
export const getRuntime = async (
  token: string,
  podName: string,
  baseUrl: string = DEFAULT_SERVICE_URLS.RUNTIMES,
): Promise<RuntimeData> => {
  validateToken(token);
  validateRequiredString(podName, 'Pod name');

  try {
    const response = await requestDatalayerAPI<any>({
      url: `${baseUrl}${API_BASE_PATHS.RUNTIMES}/runtimes/${podName}`,
      method: 'GET',
      token,
    });

    // The API returns { success: true, message: string, runtime: Runtime }
    // (Previously used 'kernel' field, now uses 'runtime')

    // Try 'runtime' field first (current API)
    if (response.runtime) {
      return {
        ...response.runtime,
        pod_name: response.runtime.pod_name || podName,
      } as RuntimeData;
    }

    // Fallback to 'kernel' field (old API)
    if (response.kernel) {
      return {
        ...response.kernel,
        pod_name: response.kernel.pod_name || podName,
      } as RuntimeData;
    }

    // Fallback if response structure is different
    return response as RuntimeData;
  } catch (error: any) {
    // Handle specific error cases
    if (error.response && error.response.status === 404) {
      // Runtime not found
      throw new Error(
        `Runtime with pod name '${podName}' not found. Please check the pod name and try again.`,
      );
    }

    // Re-throw the original error for other cases
    throw error;
  }
};

/**
 * Delete a runtime instance.
 * @param token - Authentication token
 * @param podName - The unique pod name of the runtime to delete
 * @param baseUrl - Base URL for the API (defaults to production Runtimes URL)
 * @returns Promise resolving when deletion is complete
 * @throws {Error} If authentication token is missing or invalid
 * @throws {Error} If pod name is missing or invalid
 * @throws {Error} With status 404 if the runtime is not found
 */
export const deleteRuntime = async (
  token: string,
  podName: string,
  baseUrl: string = DEFAULT_SERVICE_URLS.RUNTIMES,
): Promise<void> => {
  validateToken(token);
  validateRequiredString(podName, 'Pod name');

  try {
    return await requestDatalayerAPI<void>({
      url: `${baseUrl}${API_BASE_PATHS.RUNTIMES}/runtimes/${podName}`,
      method: 'DELETE',
      token,
    });
  } catch (error: any) {
    // Handle specific error cases
    if (error.response && error.response.status === 404) {
      // Runtime not found
      throw new Error(
        `Runtime with pod name '${podName}' not found. Cannot delete a non-existent runtime.`,
      );
    }

    // Re-throw the original error for other cases
    throw error;
  }
};

/**
 * Update a runtime instance.
 * @param token - Authentication token
 * @param podName - The unique pod name of the runtime
 * @param from - The source to update from
 * @param baseUrl - Base URL for the API (defaults to production Runtimes URL)
 * @returns Promise resolving to updated runtime details
 * @throws {Error} If authentication token is missing or invalid
 * @throws {Error} If pod name is missing or invalid
 * @throws {Error} With status 404 if the runtime is not found
 */
export const updateRuntime = async (
  token: string,
  podName: string,
  from: string,
  baseUrl: string = DEFAULT_SERVICE_URLS.RUNTIMES,
): Promise<RuntimeData> => {
  validateToken(token);
  validateRequiredString(podName, 'Pod name');

  try {
    return await requestDatalayerAPI<RuntimeData>({
      url: `${baseUrl}${API_BASE_PATHS.RUNTIMES}/runtimes/${podName}`,
      method: 'PUT',
      token,
      body: { from },
    });
  } catch (error: any) {
    // Handle specific error cases
    if (error.response && error.response.status === 404) {
      // Runtime not found
      throw new Error(
        `Runtime with pod name '${podName}' not found. Cannot update a non-existent runtime.`,
      );
    }

    // Re-throw the original error for other cases
    throw error;
  }
};

/**
 * Response from the pause/resume runtime endpoints.
 * The server returns 202 Accepted and provides a checkpoint UID
 * that can be polled for status updates.
 */
export interface PauseResumeResponse {
  success: boolean;
  message: string;
  checkpoint_id: string | null;
}

/**
 * Optional body for the pause endpoint.
 * Metadata is stored in the checkpoint Solr record created by the backend.
 */
export interface PauseRuntimeBody {
  /** Human-readable checkpoint name */
  name?: string;
  /** Checkpoint description */
  description?: string;
  /** Agent spec identifier */
  agent_spec_id?: string;
  /** Full agent spec payload to persist with the checkpoint */
  agentspec?: Record<string, any>;
  /** Additional metadata */
  metadata?: Record<string, any>;
}

/**
 * Pause a runtime by creating a CRIU checkpoint (async).
 *
 * Returns immediately with a 202 Accepted response.  The actual checkpoint
 * process runs in the background.  Use the returned `checkpoint_id` to
 * poll for status via the runtime-checkpoints API.
 *
 * @param token - Authentication token
 * @param podName - The unique pod name of the runtime to pause
 * @param baseUrl - Base URL for the API (defaults to production Runtimes URL)
 * @param body - Optional metadata to store in the checkpoint record
 * @returns Promise resolving with the checkpoint ID for status tracking
 * @throws {Error} If authentication token is missing or invalid
 * @throws {Error} If pod name is missing or invalid
 */
export const pauseRuntime = async (
  token: string,
  podName: string,
  baseUrl: string = DEFAULT_SERVICE_URLS.RUNTIMES,
  body?: PauseRuntimeBody,
): Promise<PauseResumeResponse> => {
  validateToken(token);
  validateRequiredString(podName, 'Pod name');

  return await requestDatalayerAPI<PauseResumeResponse>({
    url: `${baseUrl}${API_BASE_PATHS.RUNTIMES}/runtimes/${encodeURIComponent(podName)}/pause`,
    method: 'POST',
    token,
    ...(body ? { body } : {}),
  });
};

/**
 * Optional body for the resume endpoint.
 * Contains information needed by the operator to restore from a CRIU checkpoint.
 */
export interface ResumeRuntimeBody {
  /** Agent spec identifier (required by the operator for restore) */
  agent_spec_id?: string;
  /** Specific checkpoint timestamp to restore from */
  checkpoint_timestamp?: string;
  /** Environment name override */
  environment_name?: string;
  /** Container image override */
  container_image?: string;
  /** Target node name */
  node_name?: string;
  /** Additional metadata */
  metadata?: Record<string, any>;
}

/**
 * Resume a paused runtime by restoring from a CRIU checkpoint (async).
 *
 * Returns immediately with a 202 Accepted response.  The actual restore
 * process runs in the background.  Use the returned `checkpoint_id` to
 * poll for status.
 *
 * @param token - Authentication token
 * @param podName - The unique pod name of the runtime to resume
 * @param baseUrl - Base URL for the API (defaults to production Runtimes URL)
 * @param body - Optional body with agent_spec_id and restore options
 * @returns Promise resolving with the checkpoint ID for status tracking
 * @throws {Error} If authentication token is missing or invalid
 * @throws {Error} If pod name is missing or invalid
 */
export const resumeRuntime = async (
  token: string,
  podName: string,
  baseUrl: string = DEFAULT_SERVICE_URLS.RUNTIMES,
  body?: ResumeRuntimeBody,
): Promise<PauseResumeResponse> => {
  validateToken(token);
  validateRequiredString(podName, 'Pod name');

  return await requestDatalayerAPI<PauseResumeResponse>({
    url: `${baseUrl}${API_BASE_PATHS.RUNTIMES}/runtimes/${encodeURIComponent(podName)}/resume`,
    method: 'POST',
    token,
    ...(body ? { body } : {}),
  });
};

File: datalayer_core/mixins/runtimes.py
# Copyright (c) 2023-2025 Datalayer, Inc.
# Distributed under the terms of the Modified BSD License.

"""Runtime management module for Datalayer Core."""

import logging
import sys
from typing import Any, Optional

from datalayer_core.utils.defaults import get_default_credits_limit

logger = logging.getLogger(__name__)


class RuntimesCreateMixin:
    """Mixin for creating a Datalayer Runtime."""

    def _create_runtime(
        self,
        environment_name: str = "python-env",
        given_name: Optional[str] = None,
        credits_limit: Optional[float] = None,
        from_snapshot_uid: Optional[str] = None,
    ) -> dict[str, Any]:
        """
        Create a Runtime with the given environment name.

        Parameters
        ----------
        environment_name : str
            Name of the environment to use.
        given_name : Optional[str]
            Custom name for the runtime.
        credits_limit : Optional[float]
            Credit limit for the runtime.
        from_snapshot_uid : Optional[str]
            UID of snapshot to create runtime from.

        Returns
        -------
        dict[str, Any]
            Response containing runtime creation details.
        """
        body = {
            "type": "notebook",
            "environment_name": environment_name,
        }

        if given_name:
            body["given_name"] = given_name

        try:
            if credits_limit is None:
                response = self._fetch(  # type: ignore
                    "{}/api/iam/v1/usage/credits".format(self.urls.iam_url),  # type: ignore
                    method="GET",
                )

                if response.status_code != 200:
                    error_msg = f"Failed to fetch credits: HTTP {response.status_code}"
                    logger.error(error_msg)
                    return {"success": False, "message": error_msg}

                try:
                    raw_credits = response.json()
                except Exception as e:
                    error_msg = f"Failed to parse credits response: {str(e)}"
                    logger.error(error_msg)
                    return {"success": False, "message": error_msg}

                if "success" in raw_credits and not raw_credits["success"]:
                    error_msg = f"Credits API returned error: {raw_credits.get('message', 'Unknown error')}"
                    logger.error(error_msg)
                    return {"success": False, "message": error_msg}

                credits_limit = get_default_credits_limit(
                    raw_credits.get("reservations", []), raw_credits.get("credits", 0)
                )
                logger.debug("Runtime will use credits limit: %.2f", credits_limit)

            if credits_limit < sys.float_info.epsilon:
                error_msg = (
                    "Credits reservation is not positive. Cannot create runtime."
                )
                logger.error(error_msg)
                return {"success": False, "message": error_msg}

            body["credits_limit"] = credits_limit  # type: ignore

            if from_snapshot_uid:
                body["from"] = from_snapshot_uid

            runtime_url = "{}/api/runtimes/v1/runtimes".format(self.urls.run_url)  # type: ignore
            logger.debug(
                "Creating runtime via %s with payload keys=%s",
                runtime_url,
                sorted(body.keys()),
            )
            logger.debug("Runtime create payload: %s", body)

            response = self._fetch(  # type: ignore
                runtime_url,
                method="POST",
                json=body,
            )

            if response.status_code not in [200, 201]:
                error_msg = f"Failed to create runtime: HTTP {response.status_code}"
                logger.error(error_msg)
                try:
                    error_details = response.json()
                    if "message" in error_details:
                        error_msg += f" - {error_details['message']}"
                        logger.error("Details: %s", error_details["message"])
                except Exception:
                    pass
                return {"success": False, "message": error_msg}

            try:
                result = response.json()
                if "success" in result and not result["success"]:
                    error_msg = f"Runtime creation failed: {result.get('message', 'Unknown error')}"
                    logger.error(error_msg)
                    return {"success": False, "message": error_msg}

                logger.debug(
                    "Runtime created successfully: %s",
                    result.get("runtime", {}).get("uid", "N/A"),
                )
                return result
            except Exception as e:
                error_msg = f"Failed to parse runtime creation response: {str(e)}"
                logger.error(error_msg)
                return {"success": False, "message": error_msg}

        except Exception as e:
            error_msg = f"Unexpected error during runtime creation: {str(e)}"
            logger.error(error_msg)
            return {"success": False, "message": error_msg}


class RuntimesListMixin:
    """Mixin for listing Datalayer runtimes."""

    def _list_runtimes(self) -> dict[str, Any]:
        """
        List all available runtimes.

        Returns
        -------
        dict
            A dictionary containing the response.
        """
        try:
            response = self._fetch(  # type: ignore
                "{}/api/runtimes/v1/runtimes".format(self.urls.run_url),  # type: ignore
            )

            if response.status_code != 200:
                error_msg = f"Failed to list runtimes: HTTP {response.status_code}"
                logger.error(error_msg)
                return {"success": False, "message": error_msg}

            try:
                result = response.json()
                if "success" in result and not result["success"]:
                    error_msg = f"List runtimes failed: {result.get('message', 'Unknown error')}"
                    logger.error(error_msg)
                    return {"success": False, "message": error_msg}

                return result
            except Exception as e:
                error_msg = f"Failed to parse runtimes list response: {str(e)}"
                logger.error(error_msg)
                return {"success": False, "message": error_msg}

        except Exception as e:
            error_msg = f"Unexpected error listing runtimes: {str(e)}"
            logger.error(error_msg)
            return {"success": False, "message": error_msg}


class RuntimesTerminateMixin:
    """
    Mixin for terminating Datalayer runtimes.
    """

    def _terminate_runtime(self, pod_name: str) -> dict[str, Any]:
        """
        Terminate a Runtime with the given kernel ID.

        Parameters
        ----------
        pod_name : str
            The pod name of the runtime to terminate.

        Returns
        -------
        dict[str, Any]
            Response containing termination status.
        """
        try:
            response = self._fetch(  # type: ignore
                "{}/api/runtimes/v1/runtimes/{}".format(self.urls.run_url, pod_name),  # type: ignore
                method="DELETE",
            )

            if response.status_code in [200, 204]:
                logger.debug("Runtime %s terminated successfully", pod_name)
                return {
                    "success": True,
                    "message": "Runtime terminated successfully.",
                }
            else:
                error_msg = f"Failed to terminate runtime: HTTP {response.status_code}"
                logger.error(error_msg)
                try:
                    error_details = response.json()
                    if "message" in error_details:
                        error_msg += f" - {error_details['message']}"
                        logger.error("Details: %s", error_details["message"])
                except Exception:
                    pass
                return {"success": False, "message": error_msg}

        except Exception as e:
            error_msg = f"Unexpected error terminating runtime {pod_name}: {str(e)}"
            logger.error(error_msg)
            return {"success": False, "message": error_msg}


class RuntimesMixin(
    RuntimesCreateMixin,
    RuntimesListMixin,
    RuntimesTerminateMixin,
):
    """
    Mixin class that provides runtime management functionality.
    """
